drivers/dai/nxp/sai.c

```diff
@@ -0,0 +1,758 @@
+	 */
+	tx_config->channelMask = 0x0;
+
+	/* TODO: for now, only MCLK1 is supported */
+	tx_config->bitClock.bclkSource = kSAI_BclkSourceMclkOption1;
+
+	/* FSYNC is asserted for tdm_slot_width BCLKs */
+	tx_config->frameSync.frameSyncWidth = bespoke->tdm_slot_width;
+
+	/* serial data common configuration */
+	tx_config->serialData.dataWord0Length = bespoke->tdm_slot_width;
+	tx_config->serialData.dataWordNLength = bespoke->tdm_slot_width;
+	tx_config->serialData.dataFirstBitShifted = bespoke->tdm_slot_width;
+	tx_config->serialData.dataWordNum = bespoke->tdm_slots;
+
+	/* clock provider configuration */
+	switch (cfg->format & DAI_FORMAT_CLOCK_PROVIDER_MASK) {
+	case DAI_CBP_CFP:
+		tx_config->masterSlave = kSAI_Slave;
+		break;
+	case DAI_CBC_CFC:
+		tx_config->masterSlave = kSAI_Master;
+		break;
+	case DAI_CBC_CFP:
+	case DAI_CBP_CFC:
+		LOG_ERR("unsupported provider configuration: %d",
+			cfg->format & DAI_FORMAT_CLOCK_PROVIDER_MASK);
+		return -ENOTSUP;
+	default:
+		LOG_ERR("invalid provider configuration: %d",
+			cfg->format & DAI_FORMAT_CLOCK_PROVIDER_MASK);
+		return -EINVAL;
+	}
+
+	LOG_DBG("SAI is in %d mode", tx_config->masterSlave);
+
+	/* protocol configuration */
+	switch (cfg->format & DAI_FORMAT_PROTOCOL_MASK) {
+	case DAI_PROTO_I2S:
+		/* BCLK is active LOW */
+		tx_config->bitClock.bclkPolarity = kSAI_PolarityActiveLow;
+		/* FSYNC is active LOW */
+		tx_config->frameSync.frameSyncPolarity = kSAI_PolarityActiveLow;
+		break;
+	case DAI_PROTO_DSP_A:
+		/* FSYNC is asserted for a single BCLK */
+		tx_config->frameSync.frameSyncWidth = 1;
+		/* BCLK is active LOW */
+		tx_config->bitClock.bclkPolarity = kSAI_PolarityActiveLow;
+		break;
+	default:
+		LOG_ERR("unsupported DAI protocol: %d",
+			cfg->format & DAI_FORMAT_PROTOCOL_MASK);
+		return -EINVAL;
+	}
+
+	LOG_DBG("SAI uses protocol: %d",
+		cfg->format & DAI_FORMAT_PROTOCOL_MASK);
+
+	/* clock inversion configuration */
+	switch (cfg->format & DAI_FORMAT_CLOCK_INVERSION_MASK) {
+	case DAI_INVERSION_IB_IF:
+		SAI_INVERT_POLARITY(tx_config->bitClock.bclkPolarity);
+		SAI_INVERT_POLARITY(tx_config->frameSync.frameSyncPolarity);
+		break;
+	case DAI_INVERSION_IB_NF:
+		SAI_INVERT_POLARITY(tx_config->bitClock.bclkPolarity);
+		break;
+	case DAI_INVERSION_NB_IF:
+		SAI_INVERT_POLARITY(tx_config->frameSync.frameSyncPolarity);
+		break;
+	case DAI_INVERSION_NB_NF:
+		/* nothing to do here */
+		break;
+	default:
+		LOG_ERR("invalid clock inversion configuration: %d",
+			cfg->format & DAI_FORMAT_CLOCK_INVERSION_MASK);
+		return -EINVAL;
+	}
+
+	LOG_DBG("FSYNC polarity: %d", tx_config->frameSync.frameSyncPolarity);
+	LOG_DBG("BCLK polarity: %d", tx_config->bitClock.bclkPolarity);
+
+	/* duplicate TX configuration */
+	memcpy(rx_config, tx_config, sizeof(sai_transceiver_t));
+
+	tx_config->serialData.dataMaskedWord = ~bespoke->tx_slots;
+	rx_config->serialData.dataMaskedWord = ~bespoke->rx_slots;
+
+	tx_config->fifo.fifoWatermark = sai_cfg->tx_fifo_watermark - 1;
```

Comment:
This is because the values from the DTS are in fact the correct (logically speaking) watermark values. You can't have a NULL watermark as that wouldn't make sense.